kamon-annotation/src/main/scala/kamon/annotation/instrumentation/BaseAnnotationInstrumentation.scala

```diff
@@ -0,0 +1,176 @@
+import kamon.metric.{ CounterKey, HistogramKey, MinMaxCounterKey, instrument }
+import kamon.trace.{ Tracer, TraceContext }
+import kamon.util.Latency
+import org.aspectj.lang.{ JoinPoint, ProceedingJoinPoint }
+import org.aspectj.lang.annotation.{ Aspect, DeclareMixin }
+import org.aspectj.lang.reflect.MethodSignature
+
+import scala.collection.concurrent.TrieMap
+
+class BaseAnnotationInstrumentation {
+
+  import EnhancedELProcessor.Syntax
+
+  @inline final def registerTime(method: Method, histograms: TrieMap[String, instrument.Histogram])(implicit evalString: StringEvaluator): Any = {
+    if (method.isAnnotationPresent(classOf[Time])) {
+      val time = method.getAnnotation(classOf[Time])
+      val name = evalString(time.name())
+      val tags = evalTags(time.tags())
+      val currentHistogram = Kamon.simpleMetrics.histogram(HistogramKey(name, tags))
+      histograms.put(methodName(method), currentHistogram)
+    }
+  }
+
+  @inline final def registerHistogram(method: Method, histograms: TrieMap[String, instrument.Histogram])(implicit eval: StringEvaluator): Unit = {
+    if (method.isAnnotationPresent(classOf[Histogram])) {
+      val histogram = method.getAnnotation(classOf[Histogram])
+      val name = eval(histogram.name())
+      val tags = evalTags(histogram.tags())
+      val dynamicRange = DynamicRange(histogram.lowestDiscernibleValue(), histogram.highestTrackableValue(), histogram.precision())
+      val currentHistogram = Kamon.simpleMetrics.histogram(HistogramKey(name, tags), dynamicRange)
+      histograms.put(methodName(method), currentHistogram)
+    }
+  }
+
+  @inline final def registerMinMaxCounter(method: Method, minMaxCounters: TrieMap[String, MinMaxCounter])(implicit eval: StringEvaluator): Unit = {
+    if (method.isAnnotationPresent(classOf[MinMaxCount])) {
+      val minMaxCount = method.getAnnotation(classOf[MinMaxCount])
+      val name = eval(minMaxCount.name())
+      val tags = evalTags(minMaxCount.tags())
+      val minMaxCounter = Kamon.simpleMetrics.minMaxCounter(MinMaxCounterKey(name, tags))
+      minMaxCounters.put(methodName(method), minMaxCounter)
+    }
+  }
+
+  @inline final def registerCounter(method: Method, counters: TrieMap[String, Counter])(implicit eval: StringEvaluator): Unit = {
+    if (method.isAnnotationPresent(classOf[Count])) {
+      val count = method.getAnnotation(classOf[Count])
+      val name = eval(count.name())
+      val tags = evalTags(count.tags())
+      val counter = Kamon.simpleMetrics.counter(CounterKey(name, tags))
+      counters.put(methodName(method), counter)
+    }
+  }
+
+  @inline final def registerTrace(method: Method, traces: TrieMap[String, TraceContext])(implicit eval: StringEvaluator): Unit = {
+    if (method.isAnnotationPresent(classOf[Trace])) {
+      val count = method.getAnnotation(classOf[Trace])
+      val name = eval(count.value())
+      val tags = evalTags(count.tags())
+      val trace = Kamon.tracer.newContext(name)
```

Comment:
Here we are registering a single TraceContext for all invocations of the `@Trace`'d method and that breaks the principle of having a single TraceContext per each separate invocation of a given functionality. Instead of doing this, we should be always creating a new TraceContext when calling `processTrace`.